fem/linearform_ext.cpp

```diff
@@ -61,6 +61,7 @@ void LinearFormExtension::Assemble()
       }
 
       // Assemble the linear form
+      Vector b(elem_restrict_lex->Height(), Device::GetDeviceTemporaryMemoryType());
```

Comment:
We need to make sure `b.UseDevice(true)` is called, either directly or as part of the above constructor under some conditions.

One option for the above `Vector` constructor is to check if the `MemoryType` argument is in the currently-configured device `MemoryClass` as returned by `Device::GetDeviceMemoryClass()`, i.e. when `MemoryClassContainsType(Device::GetDeviceMemoryClass(), mt)` is true, and in that case call `UseDevice(true)`.

Reply:
Good point. I think that's already handled here (line 972 below): 

https://github.com/mfem/mfem/blob/40f6b5658e9bae7a67b0dfb8841d5b47eba4c6db/general/mem_manager.hpp#L962-L973

Does that look OK to you?

Reply:
The idea for the `USE_DEVICE` was that it is independent of the `Memory` class -- no methods in the class use the value or depend on the value, except the constructors (initialize to false) and the two methods that get and set the value.

I'm leaning toward keeping the current approach -- call `b.UseDevice(true)`. Modifying the `Vector` constructor as described above can have some undesired effects, e.g. creating a vector with the current host memory type will set the device flag when using the OpenMP backend (in some cases this is what one may want but in other cases it may not be).

Reply:
How about the above, simply setting the device flag if `IsDeviceMemory(mt))` is true?

Is there a situation when a user will specifically construct a vector with a device memory type, but not intend for vector operations to be performed on device?

Reply:
In this context, we want to have code that will set the device execution flag for OpenMP too, so using `IsDeviceMemory(mt))` is not sufficient.

Reply:
Ah yeah, I see. Maybe we should have a constructor for `Vector` that both sets the memory type to `Device::GetDeviceTemporaryMemoryType()` and sets `Vector::UseDevice(true)`.

Both as a convenience thing, and to avoid forgetting to call `UseDevice(true)` and then getting a performance penalty.